File: tests/cpu/religion/majority.test.ts
import { describe, it, expect } from 'vitest';
import { followersOf, followedReligionOf, ATHEISM_PRESSURE_PER_POP, FOUNDER_BELIEFS } from '../../../cpu/data/religion';
import { makeMap, makeState, tileAtCoords, settleAt } from '../helpers';
import { emptySeat, majorityReligionOf, dominantReligionOf, BARB_SEAT } from '../../../cpu/core/seats';
import { rosterCS, conquistadorConvert } from '../../../cpu/core/combat';
import { sameReligionToken, founderBeliefOf } from '../../../cpu/core/effects';
import { placeCityStateAt } from '../../../cpu/core/cityStates';
import { spawnUnit } from '../../../cpu/core/units';
import { CIV_LEADERS } from '../../../cpu/data/seats';
import type { City, GameState } from '../../../cpu/core/types';

// A city's MAJORITY religion, measured live (lab 2 scene E, eight draws
// including the decider; tools/civ6lab/runs/religion_20260920T183900Z.jsonl).
// Religion ids here follow the game's order: CATH < PROT < BUD. The rows pass
// the UNCONVERTED pressure the game reported (an accumulator the live game
// keeps — 300 at pop 2 after a shrink — where this engine derives 50 × pop).
const CATH = 0;
const PROT = 1;
const BUD = 2;

describe('followers — pop × pressure share, rounded, forced to sum to pop', () => {
  it('NGAZARGAMU pop 9: CATH 1, PROT 2, BUD 4, none 2 (the largest-remainder allocation)', () => {
    expect(followersOf([212, 641, 940], 9, 400)).toEqual([1, 2, 4, 2]);
  });
  it('STAVANGER pop 6, one Catholic spread: CATH 3, none 3', () => {
    expect(followersOf([220], 6, 300)).toEqual([3, 3]);
  });
  it('STAVANGER pop 3, three groups: one citizen each', () => {
    expect(followersOf([359, 710], 3, 300)).toEqual([1, 1, 1]);
  });
  it('the engine derives the unconverted pressure at 50 per citizen; pop 0 seats nobody', () => {
    expect(ATHEISM_PRESSURE_PER_POP).toBe(50);
    expect(followersOf([], 4)).toEqual([4]);
    expect(followersOf([100], 2)).toEqual([1, 1]);   // 100 vs 100: the remainder tie goes to the lower id
    expect(followersOf([100], 0)).toEqual([0, 0]);
  });
});

describe('the majority — most followers, ties by pressure, at least half the citizens', () => {
  it('NGAZARGAMU: BUD leads on both counts and still fails the half-gate (2 × 4 < 9)', () => {
    expect(followedReligionOf([212, 641, 940], 9, 400)).toBe(-1);
  });
  it('STAVANGER draw 1, pop 6: a 3-3 tie with the unconverted goes to their pressure — no majority', () => {
    expect(followedReligionOf([220], 6, 300)).toBe(-1);
  });
  it('STAVANGER draw 3, pop 2: the same shape with the pressure order flipped — Protestantism', () => {
    expect(followedReligionOf([0, 880], 2, 300)).toBe(PROT);
  });
  it('STAVANGER draw 4, pop 3: the pressure winner is a religion, yet 2 × 1 < 3', () => {
    expect(followedReligionOf([359, 710], 3, 300)).toBe(-1);
  });
  it('STAVANGER draws 5 and 6, pop 2: a 1-1 tie between religions goes to the higher PRESSURE, not the lower id', () => {
    expect(followedReligionOf([579, 532], 2, 0)).toBe(CATH);
    expect(followedReligionOf([434, 752], 2, 0)).toBe(PROT);   // the decider
  });
  it('through the engine\'s own unconverted term: a religion holding every citizen of a one-pop city is its majority; pop 0 is nobody\'s', () => {
    expect(followedReligionOf([0, 0, 300], 1)).toBe(BUD);
    expect(followedReligionOf([300], 0)).toBe(-1);
    // exactly half the citizens is enough; the unconverted at equal followers and higher pressure is not
    expect(followedReligionOf([100], 2)).toBe(CATH);        // 1-1, 100 vs 100: the lower id
    expect(followedReligionOf([90], 2)).toBe(-1);           // 1-1, 90 vs 100: the unconverted
  });
});

// ── THE SEAT'S MAJORITY (GetReligionInMajorityOfCities, measured): strictly
// more than half of the player's cities share a majority; a city with none
// counts in the denominator; a seat that founded nothing can hold one. ONE
// composer (`majorityReligionOf`) for the four readers below.
const leaderRow = (leader: string) => CIV_LEADERS.findIndex((l) => l.leader === leader);
const seatRow = (civ: string) => CIV_LEADERS.findIndex((l) => l.civ === civ);

function twoSeats(row0: number): GameState {
  const state = makeState(makeMap(20, 20, 'GRASSLAND'));
  state.seats.push(emptySeat(1));
  state.seats[0].civ = row0;
  state.seats[1].civ = seatRow('AMERICA');
  state.unitsMode = true;
  return state;
}

/** cities for `seat`, following `follow` (null = no majority) */
function cities(state: GameState, seat: number, follow: (number | null)[]): City[] {
  const spots = seat === 0 ? [[3, 3], [3, 9], [3, 15]] : [[15, 3], [15, 9], [15, 15]];
  return follow.map((g, i) => {
    const c = settleAt(state, tileAtCoords(state.map, spots[i][0], spots[i][1]).index, seat);
    c.followedReligion = g;
    return c;
  });
}

describe('the SEAT majority and its four readers', () => {
  it('more than half of the cities, the majority-less ones in the denominator; a minor by its one city; nobody for the barbarians', () => {
    const state = twoSeats(seatRow('AMERICA'));
    cities(state, 0, [1, 1, null]);
    expect(dominantReligionOf(state.seats[0])).toBe(1);
    expect(majorityReligionOf(state, 0)).toBe(1);
    state.seats[0].cities[1].followedReligion = null;          // 1 of 3
    expect(majorityReligionOf(state, 0)).toBe(-1);
    state.seats[0].cities[1].followedReligion = 0;             // 1-1-none
    expect(majorityReligionOf(state, 0)).toBe(-1);
    expect(majorityReligionOf(state, 1)).toBe(-1);             // no cities at all
    const cs = placeCityStateAt(state, 0, 'CS0', 'religious', tileAtCoords(state.map, 9, 9).index);
    cs.religionPressure = [0, 400];                            // pop 3: 2 followers of 1, 1 unconverted
    expect(majorityReligionOf(state, cs.seat)).toBe(1);
    cs.religionPressure = [0, 100];                            // 100 vs the engine's 150: 1-2, the unconverted
    expect(majorityReligionOf(state, cs.seat)).toBe(-1);
    expect(majorityReligionOf(state, BARB_SEAT)).toBe(-1);
    expect(majorityReligionOf(state, 7)).toBe(-1);
  });

  it('El Escorial: +5 against a player of ANOTHER majority religion — a major or a minor — and nothing when either side has none', () => {
    const state = twoSeats(leaderRow('PHILIP_II'));
    cities(state, 0, [0, 0, null]);
    cities(state, 1, [1, 1, null]);
    const u = spawnUnit(state, 'WARRIOR', tileAtCoords(state.map, 9, 3).index, 0)!;
    expect(rosterCS(state, u, 1, 100, false)).toBe(5);
    const cs = placeCityStateAt(state, 0, 'CS0', 'religious', tileAtCoords(state.map, 9, 12).index);
    cs.religionPressure = [0, 400];
    expect(rosterCS(state, u, cs.seat, 100, false)).toBe(5);
    cs.religionPressure = [400];                                // the minor follows Spain's own
    expect(rosterCS(state, u, cs.seat, 100, false)).toBe(0);
    for (const c of state.seats[1].cities) c.followedReligion = 0;   // the foe shares it
    expect(rosterCS(state, u, 1, 100, false)).toBe(0);
    for (const c of state.seats[1].cities) c.followedReligion = 1;
    state.seats[0].cities[1].followedReligion = null;          // Spain itself has none
    expect(rosterCS(state, u, 1, 100, false)).toBe(0);
    // another leader — Tamar, who holds no combat-strength row (America's
    // Roosevelt Corollary would pay its own +5 on the home continent)
    const plain = twoSeats(leaderRow('TAMAR'));
    cities(plain, 0, [0, 0, null]);
    cities(plain, 1, [1, 1, null]);
    const w = spawnUnit(plain, 'WARRIOR', tileAtCoords(plain.map, 9, 3).index, 0)!;
    expect(rosterCS(plain, w, 1, 100, false)).toBe(0);
  });

  it("Tamar's duplicate token: one more envoy to a minor whose city follows Georgia's majority", () => {
    const state = twoSeats(leaderRow('TAMAR'));
    cities(state, 0, [0, 0, null]);
    const cs = placeCityStateAt(state, 0, 'CS0', 'religious', tileAtCoords(state.map, 9, 9).index);
    cs.religionPressure = [400];
    expect(sameReligionToken(state, cs, 0)).toBe(1);
    cs.religionPressure = [0, 400];
    expect(sameReligionToken(state, cs, 0)).toBe(0);
    cs.religionPressure = [400];
    state.seats[0].cities[1].followedReligion = null;          // Georgia has no majority
    expect(sameReligionToken(state, cs, 0)).toBe(0);
    const plain = twoSeats(seatRow('AMERICA'));
    cities(plain, 0, [0, 0, null]);
    const cs2 = placeCityStateAt(plain, 0, 'CS0', 'religious', tileAtCoords(plain.map, 9, 9).index);
    cs2.religionPressure = [400];
    expect(sameReligionToken(plain, cs2, 0)).toBe(0);
  });

  it('Mvemba is paid the founder belief of the religion most of his cities follow; everyone else only their own', () => {
    const belief = Object.keys(FOUNDER_BELIEFS)[0];
    const state = twoSeats(leaderRow('MVEMBA'));
    state.seats[1].religion.founded = true;
    state.seats[1].religion.founder = belief;
    cities(state, 0, [1, 1, null]);
    cities(state, 1, [1, 1, 1]);
    expect(founderBeliefOf(state, 1)).toBe(belief);
    expect(founderBeliefOf(state, 0)).toBe(belief);
    state.seats[0].cities[1].followedReligion = null;
    expect(founderBeliefOf(state, 0)).toBeNull();
    const plain = twoSeats(seatRow('AMERICA'));
    plain.seats[1].religion.founded = true;
    plain.seats[1].religion.founder = belief;
    cities(plain, 0, [1, 1, null]);
    expect(founderBeliefOf(plain, 0)).toBeNull();
  });

  it("the Conquistador converts a captured city to its player's MAJORITY — a religion Spain never founded — and nothing without one", () => {
    const state = twoSeats(leaderRow('PHILIP_II'));
    cities(state, 0, [1, 1, null]);
    const [city] = cities(state, 1, [null]);
    const u = spawnUnit(state, 'CONQUISTADOR', city.centerIndex, 0)!;
    expect(state.seats[0].religion.founded).toBe(false);
    conquistadorConvert(state, u, city);
    expect(city.followedReligion).toBe(1);
    city.followedReligion = null;
    state.seats[0].cities[1].followedReligion = null;
    conquistadorConvert(state, u, city);
    expect(city.followedReligion).toBeNull();
  });
});

File: tests/gpu/majority_test.py
"""A city's MAJORITY religion — the GPU twin of `followersOf` / `followedReligionOf`,
pinned on the live-game rows of lab 2 scene E (followers are pop × pressure share by
largest remainder; the majority is the group with the most followers, ties by
pressure, at least half the citizens; the unconverted winning = none). The rows pass
the UNCONVERTED pressure the game reported (an accumulator the live game keeps —
300 at pop 2 after a shrink — where this engine derives 50 × pop).

    python tests/gpu/majority_test.py
"""

from __future__ import annotations

import sys
from pathlib import Path

import torch

sys.path.insert(0, str(Path(__file__).resolve().parent.parent.parent / "gpu"))
sys.path.insert(0, str(Path(__file__).resolve().parent))

from core import BatchSim, load_rules, load_fixture, fixture_paths  # noqa: E402
from civ_abilities_test import play, place, UNITS  # noqa: E402 — the roster seating and a unit

CATH, PROT, BUD = 0, 1, 2


def main() -> None:
    rules = load_rules()
    sim = BatchSim([load_fixture(fixture_paths()[0])], rules, device="cpu", dtype=torch.float64)
    assert int(sim._atheism_per_pop) == 50

    def T(x, dtype=torch.long):
        return torch.tensor([x], dtype=dtype)

    def followers(pres, pop, none=None):
        f, _ = sim._followers_of(T(pres), T(pop), None if none is None else T(none))
        return f[0].tolist()

    def majority(pres, pop, none=None):
        return int(sim._followed_religion(T(pres), T(pop), None if none is None else T(none))[0])

    # followers — the largest-remainder allocation
    assert followers([212, 641, 940], 9, 400) == [1, 2, 4, 2], followers([212, 641, 940], 9, 400)
    assert followers([220], 6, 300) == [3, 3]
    assert followers([359, 710], 3, 300) == [1, 1, 1]
    assert followers([100], 2) == [1, 1], "100 vs the engine's 100: the remainder tie goes to the lower id"
    assert followers([100], 0) == [0, 0]
    print("  1 followers OK — NGAZARGAMU 1/2/4/2, STAVANGER 3/3 and 1/1/1")

    # the majority
    assert majority([212, 641, 940], 9, 400) == -1, "BUD leads on both counts and fails the half-gate"
    assert majority([220], 6, 300) == -1, "a 3-3 tie with the unconverted goes to their pressure"
    assert majority([0, 880], 2, 300) == PROT
    assert majority([359, 710], 3, 300) == -1, "the pressure winner is a religion, yet 2 x 1 < 3"
    assert majority([579, 532], 2, 0) == CATH
    assert majority([434, 752], 2, 0) == PROT, "the decider: the tie goes to the higher pressure, not the lower id"
    assert majority([0, 0, 300], 1) == BUD
    assert majority([300], 0) == -1
    assert majority([100], 2) == CATH and majority([90], 2) == -1, "1-1 against the engine's 100: pressure, then the lower id"
    print("  2 majority OK — the measured rows, the decider included")

    # a batch of rows answers each row on its own, and through a leading city axis
    pres = torch.tensor([[212, 641, 940], [0, 880, 0], [434, 752, 0]], dtype=torch.long)
    pop = torch.tensor([9, 2, 2], dtype=torch.long)
    none = torch.tensor([400, 300, 0], dtype=torch.long)
    assert sim._followed_religion(pres, pop, none).tolist() == [-1, PROT, PROT]
    assert sim._followed_religion(pres.unsqueeze(0), pop.unsqueeze(0), none.unsqueeze(0)).tolist() == [[-1, PROT, PROT]]
    assert sim._followed_religion(pres, pop).tolist() == [-1, PROT, PROT], "the engine's own unconverted term agrees on these rows"
    print("  3 batched shapes OK")

    # ── THE SEAT'S MAJORITY (GetReligionInMajorityOfCities): more than half of
    # the cities, the majority-less ones in the denominator — ONE composer
    # (`_seat_majority_religion`) for the four readers below
    B0, M = 0, sim.n_majors
    assert sim.S > 0, "the fixture seats no city-state"
    s0 = 0
    sim.city_alive[B0, 0, :3] = True
    sim.city_followed[B0, 0, :3] = torch.tensor([1, 1, -1])
    assert int(sim._dominant_religion()[B0, 0]) == 1
    sim.city_followed[B0, 0, 1] = -1                       # an in-place write moves the memo's stamp
    assert int(sim._dominant_religion()[B0, 0]) == -1, "1 of 3 is not a majority"
    sim.city_followed[B0, 0, :3] = torch.tensor([1, 0, -1])
    assert int(sim._dominant_religion()[B0, 0]) == -1, "1-1-none"
    sim.city_followed[B0, 0, :3] = torch.tensor([1, 1, -1])
    sim.citystate_pop[B0, s0] = 3
    sim.city_pressure[B0, M + s0, 0, :] = 0
    sim.city_pressure[B0, M + s0, 0, 1] = 400              # 2 followers of 1, 1 unconverted
    assert int(sim._minor_followed()[B0, s0]) == 1
    seats = torch.tensor([[0, 1, 100 + s0, 200, 300, -1]])
    assert sim._seat_majority_religion(seats)[B0].tolist() == [1, -1, 1, -1, -1, -1], \
        sim._seat_majority_religion(seats)[B0].tolist()
    sim.city_pressure[B0, M + s0, 0, 1] = 100              # 100 vs the engine's 150: the unconverted
    assert int(sim._seat_majority_religion(torch.tensor([[100 + s0]]))[B0, 0]) == -1
    print("  4 seat majority OK — a major's cities, a minor's one city, nobody for the barbarians and the Free row")

    # El Escorial: +5 against a player of ANOTHER majority religion
    play(sim, 0, "SPAIN")
    W = UNITS.index("WARRIOR")
    tile = int((~sim.water[B0]).long().argmax())
    sim.city_followed[B0, 0, :3] = torch.tensor([0, 0, -1])
    sim.city_alive[B0, 1, :3] = True
    sim.city_followed[B0, 1, :3] = torch.tensor([1, 1, -1])
    sim.city_pressure[B0, M + s0, 0, 1] = 400

    def cs(foe):
        return int(sim._roster_cs(torch.tensor([[0]]), torch.tensor([[W]]), torch.tensor([[tile]]),
                                  torch.tensor([[foe]]), torch.tensor([[100]]), False)[B0, 0])
    assert cs(1) == 5, cs(1)
    assert cs(100 + s0) == 5, "a minor of another religion"
    sim.city_pressure[B0, M + s0, 0, :] = 0
    sim.city_pressure[B0, M + s0, 0, 0] = 400
    assert cs(100 + s0) == 0, "the minor follows Spain's own"
    sim.city_followed[B0, 1, :3] = torch.tensor([0, 0, -1])
    assert cs(1) == 0, "the foe shares it"
    sim.city_followed[B0, 1, :3] = torch.tensor([1, 1, -1])
    sim.city_followed[B0, 0, 1] = -1
    assert cs(1) == 0, "Spain itself has none"
    sim.city_followed[B0, 0, 1] = 0
    play(sim, 0, "AMERICA")
    assert cs(1) == 0, "another leader"
    print("  5 El Escorial OK — +5 against a major or a minor of another religion, nothing when either side has none")

    # Tamar's duplicate token at the SEND
    play(sim, 0, "GEORGIA")
    sim.city_pressure[B0, M + s0, 0, :] = 0
    sim.city_pressure[B0, M + s0, 0, 0] = 400             # the minor follows Georgia's majority (0)
    sim.seat_ext[B0, 0] = True
    sim.citystate_alive[B0, s0] = True
    sim.seat_citystate_met[B0, 0, s0] = True
    active = torch.ones(sim.B, dtype=torch.bool)

    def send() -> int:
        sim.civ_envoys_avail[B0, 0] = 1
        before = int(sim.seat_citystate_envoys[B0, 0, s0])
        sim._driven_envoys[0] = torch.tensor([s0])
        sim._seat_record_apply(0, active)
        return int(sim.seat_citystate_envoys[B0, 0, s0]) - before
    base = send()   # the League's first-envoy double may ride the first one
    sim.city_pressure[B0, M + s0, 0, :] = 0
    sim.city_pressure[B0, M + s0, 0, 1] = 400
    assert send() == 1, "the minor follows another religion"
    sim.city_pressure[B0, M + s0, 0, :] = 0
    sim.city_pressure[B0, M + s0, 0, 0] = 400
    assert send() == 2, "the envoy counts as two"
    assert base >= 2
    sim.city_followed[B0, 0, 1] = -1
    assert send() == 1, "Georgia has no majority"
    sim.city_followed[B0, 0, 1] = 0
    play(sim, 0, "AMERICA")
    assert send() == 1, "another leader"
    print("  6 Tamar OK — one more envoy to a minor of Georgia's majority religion")

    # Mvemba's borrowed founder belief
    play(sim, 0, "KONGO")
    sim.civ_founder[B0, 1] = 0
    sim.civ_religion_done[B0, 1] = True
    sim.city_followed[B0, 0, :3] = torch.tensor([1, 1, -1])
    assert int(sim._eff_founder(1)[B0]) == 0, "the founder's own"
    assert int(sim._eff_founder(0)[B0]) == 0, "Mvemba borrows it"
    assert sim._bel_stamp() != sim._bel_version, "the belief memos read the majority's planes while Kongo plays"
    if sim._bel_any:
        assert sim._seat_has_beliefs(0), "the borrowed belief counts like a claim"
    sim.city_followed[B0, 0, 1] = -1
    assert int(sim._eff_founder(0)[B0]) == -1, "no majority, nothing borrowed"
    sim.city_followed[B0, 0, 1] = 1
    play(sim, 0, "AMERICA")
    assert int(sim._eff_founder(0)[B0]) == -1, "another leader follows without borrowing"
    assert sim._bel_stamp() == sim._bel_version
    print("  7 Mvemba OK — the majority religion's founder belief, refreshed by the plane's own write counter")

    # the Conquistador converts to the captor's MAJORITY, a religion it never founded
    play(sim, 0, "SPAIN")
    C = UNITS.index("CONQUISTADOR")
    assert bool(sim._type_capture_converts[C])
    ctr = tile
    sim.city_center[B0, 0, 0] = ctr
    sim.city_followed[B0, 0, :3] = torch.tensor([-1, 1, 1])
    place(sim, ctr, C, 0)
    assert not bool(sim.civ_religion_done[B0, 0])
    sim._conquistador_convert(B0, 0, ctr)
    assert int(sim.city_followed[B0, 0, 0]) == 1, int(sim.city_followed[B0, 0, 0])
    sim.city_followed[B0, 0, :3] = torch.tensor([-1, 1, -1])
    sim._conquistador_convert(B0, 0, ctr)
    assert int(sim.city_followed[B0, 0, 0]) == -1, "no majority converts nothing"
    print("  8 Conquistador OK — converts to the captor's majority, nothing without one")
    print("BATTERY OK majority")


if __name__ == "__main__":
    main()
